foyer/general_forcefield.py

```diff
@@ -0,0 +1,417 @@
+        top : gmso.Topology
+            Molecular Topology to find atom types for.
+        use_residue_map : bool, optional, default=True
+            Spped up options for duplicates subtopology.
+        """
+
+        if use_residue_map:
+            # Detect duplicates subtopology/residues
+            # (do matching by name, assert same number
+            # of atoms)
+            typemap = find_atomtypes(top, forcefield=self)
+        else:
+            typemap = find_atomtypes(top, forcefield=self)
+
+        return typemap
+
+    def _parametrize(self, top=None,
+                        typemap=dict(),
+                        references_file=None,
+                        assert_bond_params=True,
+                        assert_angle_params=True,
+                        assert_dihedral_params=True,
+                        assert_improper_params=True,
+                        verbose=False,
+                        **kwargs):
+        """Parametrize the Topology from the typemap provided
+
+        Assign AtomTypes and BondTypes to Atoms and Bonds, respectively.
+        Creat Angles, Dihedrals, Impropers and assing corresponding
+        AngleType, DihedralTypes, and ImproperTypes.
+        Parameters
+        ----------
+
+        Returns
+        -------
+        """
+        # Generate missing angles a
+        top.identify_connections()
+
+        # Assign AtomTypes
+        for atom in top.sites:
+            atom.atom_type = self.ff.atom_types.get(typemap[top.get_index(atom)]['atomtype'])
+        if not all([a.atom_type for a in top.sites]):
+            raise ValueError('Not all atoms in topology have atom types')
+
+        # Assign BondTypes
+        for bond in top.bonds:
+            # Still need to deal with permutation
+            member_types = sorted([atom.atom_type.name for atom in bond.connection_members])
+            btype_name = '~'.join(member_types)
+            bond.bond_type = self.ff.bond_types.get(btype_name)
+
+        # Assign AngleTypes
+        for angle in top.angles:
+            # Still need to deal with permutation
+            member_types = [atom.atom_type.name for atom in angle.connection_members]
+            agtype_name = '~'.join(member_types)
+            angle.angle_type = self.ff.angle_types.get(agtype_name)
+
+        # Assign DihedralTypes
+        for dihedral in top.dihedrals:
+            # Still need to deal with permutation
+            member_types = [atom.atom_type.name for atom in dihedral.connection_members]
+            dtype_name = '~'.join(member_types)
+            dihedral.dihedral_type = self.ff.dihedral_types.get(dtype_name)
+
+        # Assing ImproperTypes
+        for improper in top.impropers:
+            # Still need to deal with permutation
+            member_types = [atom.atom_type.name for atom in improper.connection_members]
+            itype_name = '~'.join(member_types)
+            improper.improper_type = self.ff.improper_types.get(itype_name)
```

Comment:
It looks like this implicitly assumes `self.ff` is a gmso FF? What if it's an omm FF?

If there ends up having to be a lot of very-similar-but-not-exactly-the-same code, I wonder if it would be cleaner to create separate classes, or just separate functions, to handle the different FF backends.

Reply:
the `self.ff` can store both gmso FF and openmm FF. i was planning to handle the slight difference between gmso FF and mm FF by an `if else` block, but having a separate function is another good option too